File: python/myoidc.py
import os
from dotenv import load_dotenv
from azure.identity import ClientSecretCredential, DefaultAzureCredential
from pymongo import MongoClient
from pymongo.auth_oidc import OIDCCallback, OIDCCallbackContext, OIDCCallbackResult

load_dotenv()

tenant_id = os.getenv("AZURE_TENANT_ID")
client_id = os.getenv("AZURE_CLIENT_ID")
client_secret = os.getenv("AZURE_CLIENT_SECRET")
uri = os.getenv("MONGODB_URI")

class MyCallback(OIDCCallback):
    def fetch(self, context: OIDCCallbackContext) -> OIDCCallbackResult:
        #not managaed identity

        credential = ClientSecretCredential( tenant_id=tenant_id, client_id=client_id, client_secret=client_secret)
        token = credential.get_token(f"{client_id}/.default").token
        return OIDCCallbackResult(access_token=token)


props = {"OIDC_CALLBACK": MyCallback()}
try:
    c = MongoClient(uri, authMechanism="MONGODB-OIDC", authMechanismProperties=props)
    result = c.sample_analytics.transactions.find_one({})
    print(result)
    c.close()
except Exception as e:
    print(e)

File: csharp/OidcMongoApp/Myoidc.cs
using System;
using System.Threading;
using System.Threading.Tasks;
using MongoDB.Driver;
using MongoDB.Driver.Authentication.Oidc; // ✅ Correct OIDC namespace
using Azure.Identity;
using Azure.Core;
using dotenv.net;
using System.Dynamic;
using System.Text.Json;

class MyOidc
{
    static async Task Main(string[] args)
    {
        // Load environment variables from .env file
        DotEnv.Load();
        var envVars = DotEnv.Read();

        var tenantId = envVars["AZURE_TENANT_ID"];
        var clientId = envVars["AZURE_CLIENT_ID"];
        var clientSecret = envVars["AZURE_CLIENT_SECRET"];
        var mongoDbUri = envVars["MONGODB_URI"]+"&authMechanism=MONGODB-OIDC&authSource=$external";

        Console.WriteLine("🔹 Tenant ID: " + tenantId);
        Console.WriteLine("🔹 Client ID: " + clientId);
        Console.WriteLine("🔹 Connecting to MongoDB...");

        try
        {
            // Create an instance of the OIDC callback class
            var oidcCallback = new MyOidcCallback(tenantId, clientId, clientSecret);

            // Configure MongoDB client settings
            var mongoClientSettings = MongoClientSettings.FromConnectionString(mongoDbUri);
            mongoClientSettings.Credential = MongoCredential.CreateOidcCredential(oidcCallback); // ✅ Pass instance

            // Initialize MongoDB client
            var client = new MongoClient(mongoClientSettings);

            // Access database and collection
            var database = client.GetDatabase("sample_analytics");
            var collection = database.GetCollection<dynamic>("transactions");

            // Fetch data from MongoDB
            var results = await collection.Find(Builders<dynamic>.Filter.Empty).FirstOrDefaultAsync();

            Console.WriteLine("✅ Successfully connected to MongoDB!");
            //return object
            Console.WriteLine(results.ToString());
        
            foreach (var document in results)
            {
                Console.WriteLine(System.Text.Json.JsonSerializer.Serialize(document, new System.Text.Json.JsonSerializerOptions { WriteIndented = true }));
            }

        }
        catch (Exception ex)
        {
            Console.WriteLine("❌ Error: " + ex.Message);
        }
    }
}

// ✅ Define a class that implements IOidcCallback correctly
class MyOidcCallback : IOidcCallback
{
    private readonly string tenantId;
    private readonly string clientId;
    private readonly string clientSecret;

    public MyOidcCallback(string tenantId, string clientId, string clientSecret)
    {
        this.tenantId = tenantId;
        this.clientId = clientId;
        this.clientSecret = clientSecret;
    }

    // ✅ Correct implementation of required method in IOidcCallback
    public MongoDB.Driver.Authentication.Oidc.OidcAccessToken GetOidcAccessToken(OidcCallbackParameters parameters  , CancellationToken cancellationToken){
        return new MongoDB.Driver.Authentication.Oidc.OidcAccessToken("token", null);
    }

    public async Task<MongoDB.Driver.Authentication.Oidc.OidcAccessToken> GetOidcAccessTokenAsync(OidcCallbackParameters parameters, CancellationToken cancellationToken)
    {
        var credential = new ClientSecretCredential(tenantId, clientId, clientSecret);
        var accessToken = await credential.GetTokenAsync(
            new TokenRequestContext(new[] { clientId+"/.default" }),
            cancellationToken
        );

        if (string.IsNullOrEmpty(accessToken.Token))
        {
            throw new Exception("❌ Failed to retrieve a valid access token.");
        }

        Console.WriteLine("✅ OIDC Token Acquired");
        Console.WriteLine("token: " + accessToken.Token);

        return new MongoDB.Driver.Authentication.Oidc.OidcAccessToken(accessToken.Token, null);
    }
}
